### PR TITLE
Bugfix for #4530: not able to test ansible operator locally

### LINKED ISSUE #4530: Not able to test Ansible operator locally or on a k3s cluster
#### What did you do?

I am trying to follow this doc:https://sdk.operatorframework.io/docs/building-operators/ansible/development-tips/ which creates ansible operator by using kubernetes collections to install configmap but when i am trying to run it locally via Ansible-runner or on cluster, i am not seeing any configmaps created

#### What did you expect to see?

configmaps get created via ansible operator

#### What did you see instead? Under which circumstances?

I am getting reconciler error when i am running make run command as shown below:
{"level":"error","ts":1613395806.9045317,"logger":"runner","msg":"Error symlinking latest artifacts","job":"7504504064263669287","name":"foo-sample","namespace":"default","error":"symlink /tmp/ansible-operator/runner/foo.example.com/v1alpha1/Foo/default/foo-sample/artifacts/7504504064263669287 /tmp/ansible-operator/runner/foo.example.com/v1alpha1/Foo/default/foo-sample/artifacts/latest: no such file or directory","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/home/travis/gopath/pkg/mod/github.com/go-logr/zapr@v0.2.0/zapr.go:132\ngithub.com/operator-framework/operator-sdk/internal/ansible/runner.(*runner).Run.func1\n\tinternal/ansible/runner/runner.go:284"}
{"level":"error","ts":1613395806.9107356,"logger":"reconciler","msg":"Failed to get ansible-runner stdout","job":"7504504064263669287","name":"foo-sample","namespace":"default","error":"open /tmp/ansible-operator/runner/foo.example.com/v1alpha1/Foo/default/foo-sample/artifacts/7504504064263669287/stdout: no such file or directory","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/home/travis/gopath/pkg/mod/github.com/go-logr/zapr@v0.2.0/zapr.go:132\ngithub.com/operator-framework/operator-sdk/internal/ansible/controller.(*AnsibleOperatorReconciler).Reconcile\n\tinternal/ansible/controller/reconcile.go:226\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/home/travis/gopath/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.0/pkg/internal/controller/controller.go:293\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/home/travis/gopath/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.0/pkg/internal/controller/controller.go:248\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1.1\n\t/home/travis/gopath/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.0/pkg/internal/controller/controller.go:211\nk8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1\n\t/home/travis/gopath/pkg/mod/k8s.io/apimachinery@v0.20.1/pkg/util/wait/wait.go:185\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1\n\t/home/travis/gopath/pkg/mod/k8s.io/apimachinery@v0.20.1/pkg/util/wait/wait.go:155\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil\n\t/home/travis/gopath/pkg/mod/k8s.io/apimachinery@v0.20.1/pkg/util/wait/wait.go:156\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/home/travis/gopath/pkg/mod/k8s.io/apimachinery@v0.20.1/pkg/util/wait/wait.go:133\nk8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext\n\t/home/travis/gopath/pkg/mod/k8s.io/apimachinery@v0.20.1/pkg/util/wait/wait.go:185\nk8s.io/apimachinery/pkg/util/wait.UntilWithContext\n\t/home/travis/gopath/pkg/mod/k8s.io/apimachinery@v0.20.1/pkg/util/wait/wait.go:99"}
{"level":"error","ts":1613395806.910875,"logger":"controller-runtime.manager.controller.foo-controller","msg":"Reconciler error","name":"foo-sample","namespace":"default","error":"open /tmp/ansible-operator/runner/foo.example.com/v1alpha1/Foo/default/foo-sample/artifacts/7504504064263669287/stdout: no such file or directory","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/home/travis/gopath/pkg/mod/github.com/go-logr/zapr@v0.2.0/zapr.go:132\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/home/travis/gopath/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.0/pkg/internal/controller/controller.go:297\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/home/travis/gopath/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.0/pkg/internal/controller/controller.go:248\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1.1\n\t/home/travis/gopath/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.0/pkg/internal/controller/controller.go:211\nk8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1\n\t/home/travis/gopath/pkg/mod/k8s.io/apimachinery@v0.20.1/pkg/util/wait/wait.go:185\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1\n\t/home/travis/gopath/pkg/mod/k8s.io/apimachinery@v0.20.1/pkg/util/wait/wait.go:155\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil\n\t/home/travis/gopath/pkg/mod/k8s.io/apimachinery@v0.20.1/pkg/util/wait/wait.go:156\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/home/travis/gopath/pkg/mod/k8s.io/apimachinery@v0.20.1/pkg/util/wait/wait.go:133\nk8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext\n\t/home/travis/gopath/pkg/mod/k8s.io/apimachinery@v0.20.1/pkg/util/wait/wait.go:185\nk8s.io/apimachinery/pkg/util/wait.UntilWithContext\n\t/home/travis/gopath/pkg/mod/k8s.io/apimachinery@v0.20.1/pkg/util/wait/wait.go:99"}

#### Environment

**Operator type:**

/language ansible 

**Kubernetes cluster type:**

k3s 1.19.4
`$ operator-sdk version`
operator-sdk version: "v1.4.0", commit: "67f9c8b888887d18cd38bb6fd85cf3cf5b94fd99", kubernetes version: "1.19.4", go version: "go1.15.5", GOOS: "linux", GOARCH: "amd64"

`$ go version` (if language is Go)

`$ kubectl version`

glcgadmin@gl1-seednode:~/foo-operator> kubectl version
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.4+k3s1", GitCommit:"2532c10faad43e2b6e728fdcc01662dc13d37764", GitTreeState:"clean", BuildDate:"2020-11-18T22:11:18Z", GoVersion:"go1.15.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.4+k3s1", GitCommit:"2532c10faad43e2b6e728fdcc01662dc13d37764", GitTreeState:"clean", BuildDate:"2020-11-18T22:11:18Z", GoVersion:"go1.15.2", Compiler:"gc", Platform:"linux/amd64"}

#### Possible Solution

Based on documentation, it should work ideally. Since, I am new to operators, i am hoping to get this issue fixed here,
#### Additional context

I have installed ansible-runner v1.4.7 and ansible-runner-http v1.0.0 and installed requirements.txt via ansible-galaxy

### PR DESCRIPTION
**Pull request by rashmigottipati**
Closes #4530

**Checklist**

If the pull request includes user-facing changes, extra documentation is required:
- [ ] Add a new changelog fragment in `changelog/fragments` (see [`changelog/fragments/00-template.yaml`](https://github.com/operator-framework/operator-sdk/tree/master/changelog/fragments/00-template.yaml))
- [ ] Add or update relevant sections of the docs website in [`website/content/en/docs`](https://github.com/operator-framework/operator-sdk/tree/master/website/content/en/docs)

Signed-off-by: rashmigottipati <chowdary.grashmi@gmail.com>

### CODE REVIEW

**Comment by rashmigottipati** [comment]
Based on the discussion from the issue #4530, I'm gating symlinking based on the existence of artifact dir and it prevents all the other errors from occurring. 
But can we always conclude that ansible runner didn't run if the artifacts directory isn't found? 

Any thoughts on this @fabianvf @estroz?

Also FYI, I've tested this PR and it shows the expected error that ansible runner was not found and doesn't show the symlink errors. Here's some output: 

`{"level":"error","ts":1621961078.7810612,"logger":"runner","msg":"Ansible-runner binary does not exist","job":"1874068156324778273","name":"memcached-sample","namespace":"default","error":"stat /tmp/ansible-operator/runner/cache.example.com/v1alpha1/Memcached/default/memcached-sample/artifacts/latest: no such file or directory","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/Users/rashmigottipati/go/pkg/mod/github.com/go-logr/zapr@v0.2.0/zapr.go:132\ngithub.com/operator-framework/operator-sdk/internal/ansible/runner.(*runner).Run.func1\n\toperator-sdk/internal/ansible/runner/runner.go:282"} `

File: internal/ansible/runner/runner.go [hunk 1]
```diff
@@ -277,6 +277,11 @@ func (r *runner) Run(ident string, u *unstructured.Unstructured, kubeconfig stri
 		// link the current run to the `latest` directory under artifacts
 		currentRun := filepath.Join(inputDir.Path, "artifacts", ident)
 		latestArtifacts := filepath.Join(inputDir.Path, "artifacts", "latest")
+		// verify if artifacts directory exists and error out if it does not exist
+		if _, err := os.Stat(latestArtifacts); os.IsNotExist(err) {
+			logger.Error(err, "Ansible-runner binary does not exist")
```
Comment (fabianvf) [inline internal/ansible/runner/runner.go:282]: Hmm, we don't know for sure that the issue is that the binary wasn't found, maybe something like

```suggestion
			logger.Error(err, "Ansible-runner does not appear to have run, the binary may be missing")
```

Comment (estroz) [reply internal/ansible/runner/runner.go:282]: You're already `Lstat`ing the artifacts path below, so you can just check for a non-exist error there and skip symlinking if true.

**Review by estroz — CHANGES_REQUESTED** [review]
There needs to be some `ansible-runner` path validation [here](https://github.com/operator-framework/operator-sdk/blob/master/internal/ansible/runner/runner.go#L130) imo:
```diff
+const ansibleRunnerBin = "ansible-runner"

// New - creates a Runner from a Watch struct
func New(watch watches.Watch, runnerArgs string) (Runner, error) {
	var path string
	var cmdFunc, finalizerCmdFunc cmdFuncType

	err := watch.Validate()
	if err != nil {
		log.Error(err, "Failed to validate watch")
		return nil, err
	}

+	if _, err := exec.LookPath(ansibleRunnerBin); err != nil {
+		return nil, err
+	}

```

File: internal/ansible/runner/runner.go [hunk 2]
```diff
@@ -277,6 +277,11 @@ func (r *runner) Run(ident string, u *unstructured.Unstructured, kubeconfig stri
 		// link the current run to the `latest` directory under artifacts
 		currentRun := filepath.Join(inputDir.Path, "artifacts", ident)
 		latestArtifacts := filepath.Join(inputDir.Path, "artifacts", "latest")
+		// verify if artifacts directory exists and error out if it does not exist
+		if _, err := os.Stat(latestArtifacts); os.IsNotExist(err) {
```
Comment (estroz) [inline internal/ansible/runner/runner.go:281]: The symlink should be attempted only if `latestArtifacts` does not exist, or it did and `os.Remove(latestArtifacts)` succeeded. This should look something like
```go
		if _, err = os.Lstat(latestArtifacts); err != nil {
			if !errors.Is(err, os.ErrNotExist) {
				logger.Error(err, "Latest artifacts dir")
				return
			}
		} else if err = os.Remove(latestArtifacts); err != nil {
			logger.Error(err, "Error removing the latest artifacts symlink")
			return
		}

		if err = os.Symlink(currentRun, latestArtifacts); err != nil {
			logger.Error(err, "Error symlinking latest artifacts")
		}
```

File: internal/ansible/runner/runner.go [hunk 3]
```diff
@@ -277,11 +283,16 @@ func (r *runner) Run(ident string, u *unstructured.Unstructured, kubeconfig stri
 		// link the current run to the `latest` directory under artifacts
 		currentRun := filepath.Join(inputDir.Path, "artifacts", ident)
 		latestArtifacts := filepath.Join(inputDir.Path, "artifacts", "latest")
-		if _, err = os.Lstat(latestArtifacts); err == nil {
-			if err = os.Remove(latestArtifacts); err != nil {
-				logger.Error(err, "Error removing the latest artifacts symlink")
+		if _, err = os.Lstat(latestArtifacts); err != nil {
+			if !errors.Is(err, os.ErrNotExist) {
+				logger.Error(err, "Latest artifacts dir doesn't exist")
```
Comment (estroz) [inline internal/ansible/runner/runner.go:288]:
```suggestion
				logger.Error(err, "Latest artifacts dir has error")
```

Comment (estroz) [inline internal/ansible/runner/runner.go:288]:
```suggestion
				logger.Error(err, "Latest artifacts dir has error")
```

**Review by estroz — APPROVED** [review]
/lgtm

**Comment by estroz** [comment]
Looks like some unit tests do not assume the environment has been set up when testing `New()`. It's probably worth moving path lookup to the top of [`Run()`](https://github.com/operator-framework/operator-sdk/pull/4944/files#diff-40a9a64aa847bb7816a7eb40cf6ec6b3ca70a7d4066e320d95fe9c0e0cb2ac11R188) for lazy environment validation.

**Pull request merged**